### PR TITLE
Nick

### PR DESCRIPTION
**Pull request by nrpriore**

### CODE REVIEW

File: Assets/Scripts/Communication/Server.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		request["pw"]		= pw;
+		Communication.SendString(Json.ToString(request));
+		string response = null;
+		while(response == null) {
+			response = Communication.RecvString();
+		}
+		bool success = (bool)Json.ToDict(response)["Success"];
+		if(success) {
+			string _loginToken = Json.ToDict(response)["Token"].ToString();
+			string _encryptedToken = AES.Encrypt(_loginToken, GeneratePasswordHash());
```
Comment (akkowal2) [inline Assets/Scripts/Communication/Server.cs]: Lets to try to just call Json.ToDict once when we get a response.  Don't want to have to unnecessarily parse the response over and over every time we want to get data from it.

File: Assets/Scripts/Communication/Server.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			PlayerPrefs.SetString("session", _encryptedToken);
+			Debug.Log("user '" + username + "' logged in with token: " + _loginToken);
+		}
+		return success;
+	}
+
+	public static bool RetryLogin() {
+		var request = new Dictionary<string, object>();
+		string _encryptedToken = PlayerPrefs.GetString("session");
+		string _loginToken = AES.Decrypt(_encryptedToken, GeneratePasswordHash());
```
Comment (akkowal2) [inline Assets/Scripts/Communication/Server.cs]: Can we name the GeneratePasswordHash to something different?  We aren't really generating a password hash here.

File: Assets/Scripts/Communication/Server.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		string response = null;
+		while (response == null) {
+			response = Communication.RecvString();
+		}
+		bool success = (bool)Json.ToDict(response)["Success"];
+		Debug.Log("user '" + username + "' created: " + success);
+		return success;
+	}
+
+	public static string GeneratePasswordHash() {
```
Comment (akkowal2) [inline Assets/Scripts/Communication/Server.cs]: Rename?  Mentioned in comment above.

**Comment by akkowal2** [comment]
General Comment @nrpriore 

Try to comment more on the different functions you are writing.  EvalToken is not very descriptive to me.  Makes it a ton harder to read and follow.  We should try to comment as we go so we don't have a ton of code that we have to go back and comment later.

File: Assets/Scripts/Communication/Server.cs (hunk 1 above)
Comment (nrpriore) [reply Assets/Scripts/Communication/Server.cs]: new var strResponse to call RecvString(). (making response 'var' doesn't work).
then setting new var response to Json.ToDict once

File: Assets/Scripts/Communication/Server.cs (hunk 2 above)
Comment (nrpriore) [reply Assets/Scripts/Communication/Server.cs]: renamed to GenerateAESKey

File: Assets/Scripts/Communication/Server.cs (hunk 3 above)
Comment (nrpriore) [reply Assets/Scripts/Communication/Server.cs]: renamed to GenerateAESKey

File: Assets/Scripts/Communication/Server.cs (hunk 1 above)
Comment (akkowal2) [reply Assets/Scripts/Communication/Server.cs]: That's fine.  I'm saying to just not call Json.ToDict over and over.  Maybe create a new dictionary variable then?  You are parsing each time you call ToDict which is not necessary.

Comment (nrpriore) [reply Assets/Scripts/Communication/Server.cs]: yeah I was just commenting the fix! Just to show via comment rather than having to look back at the code

Comment (nrpriore) [reply Assets/Scripts/Communication/Server.cs]: string strResponse = null;
while(strResponse == null) {
	strResponse = Communication.RecvString();
}
var response = Json.ToDict(strResponse);

Comment (akkowal2) [reply Assets/Scripts/Communication/Server.cs]: Ah okay, didn't know.

File: Assets/Scripts/Communication/Server.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		}
+		return success;
+	}
+
+	// Used to login to server with cached session token
+	public static bool RetryLogin() {
+		// Create the request, decrypt session token, and send it
+		var request = new Dictionary<string, object>();
+		string _encryptedToken = PlayerPrefs.GetString("session");
+		string _loginToken = AES.Decrypt(_encryptedToken, GenerateAESKey());
```
Comment (akkowal2) [inline Assets/Scripts/Communication/Server.cs:65]: Probably need a check here to see if we are able to get the session token successfully from playerprefs.  If it doesn't exist there, we should fall back to the normal login with username and password.  That can come later, just wanted to point that out.

File: Assets/Scripts/GameController.cs [hunk 5]
```diff
@@ -48,105 +52,238 @@ public class GameController : MonoBehaviour {
...
+		}else if(action == "move") {
+			// If there is no range remaining (i.e. the unit can NOT move into this token)
+			// and the action is currently "move"
+			if(prevToken.CurrentUnit != null) {
+				// We can't move here (remainingrange is less than 0)
+				// We also can't move to the previous token because prevToken.CurrentUnit != null
+				// So we need to evaluate the tokens surrounding the PREVIOUS token
+				// Use AtkRange-1 because the previous token uses the first tick of attack range
+				// If we saved the previous previous token, we could eval surrounding with AtkRange
+				EvalSurroundingTokenActions(unit, prevToken, unit.GetStat("AttackRange").Value-1, "attack");
```
Comment (akkowal2) [inline Assets/Scripts/GameController.cs:166]: General Comment: For commenting code, use /* */ when you are making multi-line comments.  (Personal Preference Opinion following): I think we should try to do most of the explaining of large parts of the code in the function comment (comments above the definition of the function).  Stuff like the above disrupts my ability to follow the code.  I would try to use 1-liner concise comments for in-line commenting.  Obviously there would be instances that require a little more explaining but 5+ lines of in-line commenting in a row is a bit excessive to me.

Comment (nrpriore) [reply Assets/Scripts/GameController.cs:166]: FINE

Comment (nrpriore) [reply Assets/Scripts/GameController.cs:166]: lol jk, but in reality, I think the description in that case is a little necessary to explain that particular check. Is that something that would go above the function?

Comment (akkowal2) [reply Assets/Scripts/GameController.cs:166]: I put the (Personal Preference Opinion) warning first for a reason.  You can argue if you disagree :)

Comment (akkowal2) [reply Assets/Scripts/GameController.cs:166]: I would just see if you can make it more concise.  For instance, "// We can't move here (remainingrange is less than 0)", I can deduce that based on the if statement.  It didn't hit the >= 0 part so maybe that part doesn't need its own dedicated comment here.  Just talking aloud.

Comment (akkowal2) [reply Assets/Scripts/GameController.cs:166]: I know I am getting a bit picky but I guess my recommendation is try to be more concise with comments and also consider if the comment you are making is really necessary.  I just don't want our code to turn into something like this.

var x = 1;  //Assigning the value 1 to x

**Review by akkowal2 — APPROVED** [review]

**Pull request merged**